apps/frontend/src/app/PageSettings/DatabaseCard.tsx

```diff
@@ -45,6 +63,156 @@ export default function DatabaseCard({
   )
 }
 
+async function createCloudFile(
+  database: ArtCharDatabase,
+  index: number,
+  fileID?: string
+) {
+  const JSONStr = JSON.stringify(database.exportGOOD())
+  if (!fileID) {
+    try {
+      const { result } = await gapi.client.drive.files.create({
+        resource: {
+          name: `genshin-optimizer-${index}.json`,
+          parents: ['appDataFolder'],
+        },
+        fields: 'id',
+      })
+      fileID = result.id
+    } catch (e) {
+      console.error('create cloud file for id failed', e)
+      // TODO: better error handling
+      return
+    }
+  }
+  try {
+    await gapi.client.request({
+      path: `/upload/drive/v3/files/${fileID}`,
+      method: 'PATCH',
+      params: { uploadType: 'media' },
+      body: JSONStr,
+    })
+  } catch (e) {
+    console.error('updating cloud file failed', e)
+    // TODO: better error handling
+  }
+}
+
+async function uploadDataToDrive(
```

Comment:
I would recommend adding in a `callback: ()=>void` parameter that invokes a callback when operation is done/errored.
Same to `downloadDataFromDrive`

Reply:
What logic I should include in the callback?

Reply:
invoke the `callback()` at 3 points:
- error
- update completed
- user cancel auth flow